### PR TITLE
Profile privacy settings

### PR DESCRIPTION
**Pull request by elp49**
8Capture
9Capture
1Capture
2Capture
3Capture
4Capture
5Capture
6Capture
7Capture

### CODE REVIEW

**Comment by jcarrete5** [comment]
~~Don't delete the `package-lock.json` file~~

**Comment by jcarrete5** [comment]
> Don't delete the `package-lock.json` file

nevermind. The diff looked weird but the file still exists

File: frontend/src/components/profile/ProfileSidebar.tsx [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+import LockIcon from "@material-ui/icons/Lock"
+
+// ProfileSidebar props.
+type ProfileSidebarProps = {
+  firstName: string
+  lastName: string
+  componentStateSetters: ProfileComponentSetters
+}
+
+// Aitem listed in the sidebar.
```
Comment (jcarrete5) [inline frontend/src/components/profile/ProfileSidebar.tsx]:
```suggestion
// An item listed in the sidebar.
```

File: api/models.py [hunk 2]
```diff
@@ -12,7 +12,43 @@ class Profile(models.Model):
...
+    )
+    interests_privacy = models.CharField(
+        max_length=2, choices=Setting.choices, default=Setting.PUBLIC
+    )
+
+
+class Interest(models.Model):
+    profile = models.ForeignKey(
+        Profile, on_delete=models.CASCADE, verbose_name="related profile"
+    )
```
Comment (jcarrete5) [inline api/models.py]: As we discussed the other day, this is probably better suited as a many-to-many field in the Profile.

File: api/views.py [hunk 3]
```diff
@@ -82,20 +84,112 @@ def post(self, request):
...
+        try:
+            # Updated user interests.
+            updatedInterests = request.data.pop("interests")
+            self.updateInterests(updatedInterests, request.user.profile)
+        except KeyError:
+            # User did not updates their interests.
+            pass
+
+        # Update profile content.
+        return self.update(request=request)
```
Comment (jcarrete5) [inline api/views.py]: This is related to the previous comment and to what we discussed the other day. This would be simpler by taking advantage of a many-to-many field on the profile model.

File: frontend/src/FullPageRoutes/Profile.tsx [hunk 4]
```diff
@@ -33,93 +34,145 @@ const useStyles = makeStyles({
...
+  let birthday = null
+  if (birth_date !== null) {
+    birthday = new Date(birth_date)
+    // Typescript removes a day.
+    birthday.setDate(birthday.getDate() + 1)
+  }
+
+  return birthday
+}
+
```
Comment (jcarrete5) [inline frontend/src/FullPageRoutes/Profile.tsx]: I'm a bit confused why we need to add 1 to the date here. Both the javascript Date and the python datetime.date appear to handle this correctly i.e. the date starts at 1. What was the problem you were having by not adding 1?

File: frontend/src/FullPageRoutes/Profile.tsx [hunk 5]
```diff
@@ -33,93 +34,145 @@ const useStyles = makeStyles({
...
+  }, [])
+
+  // Component state hooks.
+  const { componentStates, componentSetters } = useProfileComponents()
+  const {
+    isProfileActive,
+    isPrivacyActive,
+    // isNotificationsActive,
+    // isSettingsActive,
+    // isContactUsActive,
```
Comment (jcarrete5) [inline frontend/src/FullPageRoutes/Profile.tsx]:
```suggestion
```

File: frontend/src/FullPageRoutes/Profile.tsx [hunk 6]
```diff
@@ -33,93 +34,145 @@ const useStyles = makeStyles({
...
+            {isNotificationsActive && (
+              // Add notifications component here
+            )}
+            {isSettingsActive && (
+              // Add settings component here
+            )}
+            {isContactUsActive && (
+              // Add contact us component here
+            )}
+          */}
```
Comment (jcarrete5) [inline frontend/src/FullPageRoutes/Profile.tsx]:
```suggestion
```

File: frontend/src/components/profile/EditableProfileContent.tsx [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+    height: 200,
+    width: 200,
+    display: "flex",
+    overflow: "hidden",
+    alignItems: "center",
+    borderRadius: "50%",
+    justifyContent: "center",
+  },
+  image: {
+    // color: "transparent",
```
Comment (jcarrete5) [inline frontend/src/components/profile/EditableProfileContent.tsx]:
```suggestion
```

File: frontend/tsconfig.json [hunk 8]
```diff
@@ -8,7 +8,9 @@
         "noImplicitThis": true,
         "isolatedModules": true,
         "allowJs": true,
-        "moduleResolution": "node"
+        "moduleResolution": "node",
+        "resolveJsonModule": true,
+        "esModuleInterop": true
```
Comment (jcarrete5) [inline frontend/tsconfig.json]: Why was this added? To import json files in the code? Modifications to the build are almost always more appropriate in the webpack config since we are using webpack. If you need to import json (which shouldn't be necessary if the interests are stored on the backend), then a json-loader in the webpack config would be more appropriate.

**Review by jcarrete5 — CHANGES_REQUESTED** [review]

File: api/models.py (hunk 2 above)
Comment (mneedleman) [reply api/models.py]: I agree

File: frontend/src/FullPageRoutes/Profile.tsx (hunk 4 above)
Comment (mneedleman) [reply frontend/src/FullPageRoutes/Profile.tsx]: See discord

File: frontend/src/FullPageRoutes/Profile.tsx [hunk 9]
```diff
@@ -33,93 +34,131 @@ const useStyles = makeStyles({
...
+              editableSettings={editableSettings}
+              privacySetters={privacySetters}
+              setPrivacySettings={setPrivacySettings}
+              userId={userId}
+            />
+          )}
+        </Grid>
       </Grid>
-    </Grid>
+    </div>
```
Comment (mneedleman) [inline frontend/src/FullPageRoutes/Profile.tsx]: Why is this wrapped in a div?

Comment (elp49) [reply frontend/src/FullPageRoutes/Profile.tsx]: good eye, this was old code

File: api/models.py (hunk 2 above)
Comment (rao53) [reply api/models.py]: ^

**Review by rao53 — COMMENTED** [review]
Ignore ......

**Review by rao53 — DISMISSED** [review]

File: frontend/src/components/profile/ProfileContent.tsx [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+                      </Grid>
+                    </Grid>
+                  ))}
+                </Grid>
+              </Grid>
+            </Grid>
+          </Paper>
+        </Grid>
+      </Grid>
+    </div>
```
Comment (mneedleman) [inline frontend/src/components/profile/ProfileContent.tsx]: Why is this wrapped in a div?

File: frontend/src/components/profile/ProfileSidebar.tsx [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+                      {title}
+                    </Typography>
+                  </Grid>
+                </Grid>
+              </Link>
+            </Grid>
+          </Grid>
+        ))}
+      </Grid>
+    </div>
```
Comment (mneedleman) [inline frontend/src/components/profile/ProfileSidebar.tsx]: Why is this wrapped in a div?

**Review by mneedleman — DISMISSED** [review]
Not too excited about the div wrappers. I'd rather they be fragments.
Also your `Object.keys.map` can be replaced by a Ramda function, but is fine for now.

File: api/migrations/0010_auto_20210402_2245.py [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+        data = json.load(fileData)
+        for interestData in data:
+            try:
+                category = interestData["category"]
+                title = interestData["title"]
+                interest = Interest(category=category, title=title)
+                interest.save()
+            except KeyError:
+                # Interest does not follow category/title structure.
+                pass
```
Comment (jcarrete5) [inline api/migrations/0010_auto_20210402_2245.py]: Log this error so we know when it happens. This will tell us if there is a mistake in the interests file. Silently ignoring errors is almost never a good idea. 
```suggestion
            except KeyError as e:
                print(f"KeyError reading interests.json: {e}")
```

File: api/models.py [hunk 13]
```diff
@@ -12,7 +12,41 @@ class Profile(models.Model):
...
+    gender_privacy = models.CharField(
+        max_length=2, choices=Setting.choices, default=Setting.PRIVATE
+    )
+    interests_privacy = models.CharField(
+        max_length=2, choices=Setting.choices, default=Setting.PUBLIC
+    )
+
+
+class Interest(models.Model):
+    profiles = models.ManyToManyField(Profile)
```
Comment (jcarrete5) [inline api/models.py:47]: When I mentioned many-to-many, I assumed it would be added to the profile model not the interests model, but I don't think it actually matters.

File: frontend/src/FullPageRoutes/Profile.tsx [hunk 14]
```diff
@@ -33,92 +34,125 @@ const useStyles = makeStyles({
   },
 })
 
+function parseBirthday(birth_date: Nullable<Date>): Nullable<Date> {
+  let birthday = null
+  if (birth_date !== null) {
+    birthday = new Date(birth_date)
+    // Typescript is zero indexed but django is not.
```
Comment (jcarrete5) [inline frontend/src/FullPageRoutes/Profile.tsx]:
```suggestion
    // We have Django configured to store dates using local time.
    // Date objects assume the date is UTC when parsing so we add 1
    // to correct for the EST/EDT - UTC timezone difference
```

**Review by jcarrete5 — CHANGES_REQUESTED** [review]

**Comment by sb3958** [comment]
It looks good.

File: api/views.py [hunk 15]
```diff
@@ -82,21 +84,70 @@ def post(self, request):
...
+
+
+class InterestRetrieveView(generics.RetrieveAPIView):
+    serializer_class = InterestSerializer
+    queryset = Interest.objects.all()
+
+    def get(self, request):
+        allInterests = Interest.objects.all()
+        data = self.serializer_class(allInterests, many=True).data
+        return response.Response(data=data, status=status.HTTP_200_OK)
```
Comment (mneedleman) [inline api/views.py]: Since you declared the queryset, you might as well use it

File: frontend/src/components/profile/EditableProfileContent.tsx [hunk 16]
```diff
@@ -0,0 +1,377 @@
...
+    // Reset editable content values.
+    setFirstName(readonlyContent.first_name)
+    setLastName(readonlyContent.last_name)
+    setBirthDate(readonlyContent.birth_date)
+    setGender(readonlyContent.gender)
+    setInterests(readonlyContent.interests)
+  }
+
+  return (
+    <div>
```
Comment (mneedleman) [inline frontend/src/components/profile/EditableProfileContent.tsx]: This can just be a fragment right?

File: frontend/src/components/profile/PrivacySettings.tsx [hunk 17]
```diff
@@ -0,0 +1,272 @@
...
+}
+
+const useStyles = makeStyles({
+  pageTitle: {
+    fontSize: "2rem",
+    fontWeight: "bold",
+  },
+  fieldsContainer: {
+    paddingTop: 40,
+    paddingBottom: 10,
```
Comment (mneedleman) [inline frontend/src/components/profile/PrivacySettings.tsx]: No unitless css please. Yes it defaults to px, but I'd prefer we be explicit. Also I'd prefer rem

File: frontend/src/components/profile/PrivacySettings.tsx [hunk 18]
```diff
@@ -0,0 +1,272 @@
...
+    // Reset editable privacy settings.
+    setFirstNamePrivacy(readonlySettings.first_name_privacy)
+    setLastNamePrivacy(readonlySettings.last_name_privacy)
+    setBirthDatePrivacy(readonlySettings.birth_date_privacy)
+    setGenderPrivacy(readonlySettings.gender_privacy)
+    setInterestsPrivacy(readonlySettings.interests_privacy)
+  }
+
+  return (
+    <div>
```
Comment (mneedleman) [inline frontend/src/components/profile/PrivacySettings.tsx]: This can just be a fragment right?

File: frontend/src/utils/profile.tsx [hunk 19]
```diff
@@ -1,81 +1,210 @@
...
 export type ProfileContentSetters = {
   setFirstName: SetStateDispatch<string>
   setLastName: SetStateDispatch<string>
-  setBirthDate: SetStateDispatch<Date>
-  setGender: SetStateDispatch<string>
-  setVideo: SetStateDispatch<Nullable<string>>
-  setId: SetStateDispatch<number>
-  setInterests: SetStateDispatch<string>
+  setBirthDate: SetStateDispatch<Date | null>
+  setGender: SetStateDispatch<GenderKey | null>
```
Comment (mneedleman) [inline frontend/src/utils/profile.tsx]: Why not use `Nullable`

**Review by mneedleman — CHANGES_REQUESTED** [review]
Some minor things I'd like to see changed

File: frontend/src/components/profile/EditableProfileContent.tsx [hunk 20]
```diff
@@ -0,0 +1,377 @@
...
+            >
+              <Grid item>
+                <Avatar src={""} className={classes.profileAvatar} />
+              </Grid>
+              <Grid item>
+                {editableContent.video ? (
+                  <video
+                    controls
+                    src={editableContent.video}
+                    className={classes.introVideo}
```
Comment (mneedleman) [inline frontend/src/components/profile/EditableProfileContent.tsx]: Instead of the raw video component, can you add a `className` to the `components/video/Video` component and use that?

File: frontend/src/components/profile/PrivacySettings.tsx (hunk 17 above)
Comment (elp49) [reply frontend/src/components/profile/PrivacySettings.tsx]: lots of pages and components currently use the default px unit. whatever we choose it should be consistent throughout the app

File: frontend/src/components/profile/EditableProfileContent.tsx (hunk 20 above)
Comment (elp49) [reply frontend/src/components/profile/EditableProfileContent.tsx]: sounds good, the RecordPage also uses a raw video should we change it there as well?

File: api/views.py (hunk 15 above)
Comment (jcarrete5) [reply api/views.py]: You don't need this get method at all. You can just delete the get method and extend `generics.ListAPIView`. This is duplicated code

File: api/views.py [hunk 21]
```diff
@@ -83,21 +85,70 @@ def post(self, request):
...
+
+class InterestRetrieveView(generics.RetrieveAPIView):
+    serializer_class = InterestSerializer
+    queryset = Interest.objects.all()
+
+    def get(self, request):
+        allInterests = self.get_queryset()
+        data = self.serializer_class(allInterests, many=True).data
+        return response.Response(data=data, status=status.HTTP_200_OK)
+
```
Comment (jcarrete5) [inline api/views.py]:
```suggestion
class InterestRetrieveView(generics.ListAPIView):
    serializer_class = InterestSerializer
    queryset = Interest.objects.all()
```

**Review by jcarrete5 — CHANGES_REQUESTED** [review]

File: api/views.py (hunk 21 above)
Comment (elp49) [reply api/views.py]: I already submitted these changes

Comment (jcarrete5) [reply api/views.py]: too fast for me.

**Review by jcarrete5 — DISMISSED** [review]

**Review by rao53 — DISMISSED** [review]

File: frontend/src/components/profile/PrivacySettings.tsx (hunk 17 above)
Comment (mneedleman) [reply frontend/src/components/profile/PrivacySettings.tsx]: We are choosing to have units on all css.

**Review by mneedleman — CHANGES_REQUESTED** [review]

File: frontend/src/components/profile/PrivacySettings.tsx (hunk 17 above)
Comment (elp49) [reply frontend/src/components/profile/PrivacySettings.tsx]: do you want to add that to this pr or create another one for it

Comment (mneedleman) [reply frontend/src/components/profile/PrivacySettings.tsx]: For the files in this PR, change it. We can fix others as they come up

**Review by mneedleman — APPROVED** [review]
